File: apps/api-rs/src/document/plain_text.rs
use super::shared::finalize_parsed;
use super::text_utils::decode_text;
use super::*;
pub(super) fn parse_plain_text(
    doc_id: Uuid,
    parse_job_id: Uuid,
    title: &str,
    bytes: &[u8],
) -> Result<ParsedDocument> {
    let text = decode_text(bytes).context("invalid_text_encoding")?;
    let mut blocks = Vec::new();
    let mut anchors = Vec::new();
    for (paragraph, start, end) in split_paragraphs_with_ranges(&text) {
        let block_id = Uuid::new_v4();
        let mut anchor = SourceAnchor::structural(
            doc_id,
            parse_job_id,
            Uuid::nil(),
            "txt",
            "paragraph",
            block_id,
            None,
            None,
            json!({"format": "txt", "index": blocks.len()}),
            &paragraph,
        );
        anchor.char_range = Some(CharRange { start, end });
        let anchor_id = anchor.anchor_id;
        anchors.push(anchor);
        blocks.push(ParsedBlock {
            block_id,
            block_index: blocks.len() as i32,
            block_type: "paragraph".to_string(),
            text: paragraph,
            heading_level: None,
            heading_path: vec![],
            page_start: None,
            page_end: None,
            slide_index: None,
            table_id: None,
            bbox: None,
            anchor_ids: vec![anchor_id],
            source_ref: json!({"format": "txt", "index": blocks.len()}),
            metadata: json!({"format": "txt"}),
        });
    }
    Ok(finalize_parsed(
        doc_id,
        parse_job_id,
        "txt",
        title,
        None,
        blocks,
        vec![],
        anchors,
    ))
}

pub(super) fn split_paragraphs(text: &str) -> Vec<String> {
    text.split("\n\n")
        .flat_map(|part| {
            let clean = part
                .lines()
                .map(str::trim)
                .filter(|line| !line.is_empty())
                .collect::<Vec<_>>()
                .join(" ");
            if clean.is_empty() {
                vec![]
            } else {
                vec![clean]
            }
        })
        .collect()
}

fn split_paragraphs_with_ranges(text: &str) -> Vec<(String, i32, i32)> {
    let mut out = Vec::new();
    let mut paragraph_start = None;
    let mut line_start = 0usize;
    for line in text.split_inclusive('\n') {
        let raw = line
            .strip_suffix('\n')
            .unwrap_or(line)
            .trim_end_matches('\r');
        if raw.trim().is_empty() {
            if let Some(start) = paragraph_start.take() {
                push_paragraph_with_range(text, &text[start..line_start], start, &mut out);
            }
        } else if paragraph_start.is_none() {
            paragraph_start = Some(line_start);
        }
        line_start += line.len();
    }
    if let Some(start) = paragraph_start {
        push_paragraph_with_range(text, &text[start..], start, &mut out);
    }
    out
}

fn push_paragraph_with_range(
    full_text: &str,
    part: &str,
    part_byte_start: usize,
    out: &mut Vec<(String, i32, i32)>,
) {
    let mut clean_lines = Vec::new();
    let mut first_line_start: Option<usize> = None;
    let mut last_line_end: Option<usize> = None;
    let mut line_byte_start = part_byte_start;
    for line in part.split_inclusive('\n') {
        let raw_line = line.strip_suffix('\n').unwrap_or(line);
        let leading = raw_line.len() - raw_line.trim_start().len();
        let trimmed = raw_line.trim();
        if !trimmed.is_empty() {
            let start = line_byte_start + leading;
            let end = start + trimmed.len();
            first_line_start.get_or_insert(start);
            last_line_end = Some(end);
            clean_lines.push(trimmed);
        }
        line_byte_start += line.len();
    }

    if clean_lines.is_empty() {
        return;
    }
    let start = first_line_start.unwrap_or(part_byte_start);
    let end = last_line_end.unwrap_or(start);
    out.push((
        clean_lines.join(" "),
        char_index_at_byte(full_text, start),
        char_index_at_byte(full_text, end),
    ));
}

pub(super) fn find_text_char_range(
    full_text: &str,
    needle: &str,
    cursor: &mut usize,
) -> Option<(i32, i32)> {
    let needle = needle.trim();
    if needle.is_empty() {
        return None;
    }
    let haystack = &full_text[*cursor..];
    let found = haystack
        .find(needle)
        .or_else(|| haystack.find(&needle.replace('\n', " ")))?;
    let start = *cursor + found;
    let end = start + needle.len();
    *cursor = end;
    Some((
        char_index_at_byte(full_text, start),
        char_index_at_byte(full_text, end),
    ))
}

fn char_index_at_byte(text: &str, byte_index: usize) -> i32 {
    text[..byte_index.min(text.len())].chars().count() as i32
}

File: apps/api-bun/src/document/plain_text.ts
// 移植自 apps/api-rs/src/document/plain_text.rs —— 行为对齐 Rust 原版，错误信息保持一致

// 纯文本解析：段落切分（含字符范围）与文本定位

import type { SourceAnchor } from '../models/source_anchor.ts';
import { sourceAnchorStructural } from '../models/source_anchor.ts';
import { newUuid } from '../infra/uuid.ts';
import { decodeText, rustLines, splitInclusiveNewline, trimRust } from './text_utils.ts';
import { finalizeParsed } from './shared.ts';
import type { ParsedBlock, ParsedDocument } from './types.ts';
import { NIL_UUID } from './types.ts';

export function parsePlainText(
  docId: string,
  parseJobId: string,
  title: string,
  bytes: Uint8Array,
): ParsedDocument {
  const text = safeDecode(bytes);
  const blocks: ParsedBlock[] = [];
  const anchors: SourceAnchor[] = [];
  for (const [paragraph, start, end] of splitParagraphsWithRanges(text)) {
    const blockId = newUuid();
    const anchor = sourceAnchorStructural(
      docId,
      parseJobId,
      NIL_UUID,
      'txt',
      'paragraph',
      blockId,
      null,
      null,
      { format: 'txt', index: blocks.length },
      paragraph,
    );
    anchor.char_range = { start, end };
    const anchorId = anchor.anchor_id;
    anchors.push(anchor);
    blocks.push({
      block_id: blockId,
      block_index: blocks.length,
      block_type: 'paragraph',
      text: paragraph,
      heading_level: null,
      heading_path: [],
      page_start: null,
      page_end: null,
      slide_index: null,
      table_id: null,
      bbox: null,
      anchor_ids: [anchorId],
      source_ref: { format: 'txt', index: blocks.length },
      metadata: { format: 'txt' },
    });
  }
  return finalizeParsed(docId, parseJobId, 'txt', title, null, blocks, [], anchors);
}

/** decode_text 的 "invalid_text_encoding" 上下文包装（对齐 Rust .context） */
function safeDecode(bytes: Uint8Array): string {
  try {
    return decodeText(bytes);
  } catch (error) {
    if (error instanceof Error) {
      throw new Error('invalid_text_encoding:' + error.message);
    }
    throw error;
  }
}

export function splitParagraphs(text: string): string[] {
  const out: string[] = [];
  for (const part of text.split('\n\n')) {
    const clean = rustLines(part)
      .map((line) => trimRust(line))
      .filter((line) => line.length > 0)
      .join(' ');
    if (clean.length > 0) out.push(clean);
  }
  return out;
}

/** 段落在 UTF-16 索引上切分，返回码点级字符范围（对齐 Rust i32 范围） */
function splitParagraphsWithRanges(text: string): Array<[string, number, number]> {
  const out: Array<[string, number, number]> = [];
  let paragraphStart: number | null = null;
  let lineStart = 0;
  for (const line of splitInclusiveNewline(text)) {
    const raw = line.endsWith('\n') ? line.slice(0, -1) : line;
    const rawNoCr = raw.endsWith('\r') ? raw.slice(0, -1) : raw;
    if (trimRust(rawNoCr).length === 0) {
      if (paragraphStart !== null) {
        pushParagraphWithRange(text, paragraphStart, lineStart, out);
        paragraphStart = null;
      }
    } else if (paragraphStart === null) {
      paragraphStart = lineStart;
    }
    lineStart += line.length;
  }
  if (paragraphStart !== null) {
    pushParagraphWithRange(text, paragraphStart, lineStart, out);
  }
  return out;
}

function pushParagraphWithRange(
  fullText: string,
  partStart: number,
  partEnd: number,
  out: Array<[string, number, number]>,
): void {
  const part = fullText.slice(partStart, partEnd);
  const cleanLines: string[] = [];
  let firstLineStart: number | null = null;
  let lastLineEnd: number | null = null;
  let lineByteStart = partStart;
  for (const line of splitInclusiveNewline(part)) {
    const rawLine = line.endsWith('\n') ? line.slice(0, -1) : line;
    const leading = rawLine.length - trimStartLen(rawLine);
    const trimmed = trimRust(rawLine);
    if (trimmed.length > 0) {
      const start = lineByteStart + leading;
      const end = start + trimmed.length;
      if (firstLineStart === null) firstLineStart = start;
      lastLineEnd = end;
      cleanLines.push(trimmed);
    }
    lineByteStart += line.length;
  }
  if (cleanLines.length === 0) return;
  const start = firstLineStart ?? partStart;
  const end = lastLineEnd ?? start;
  out.push([
    cleanLines.join(' '),
    charIndexAt(fullText, start),
    charIndexAt(fullText, end),
  ]);
}

function trimStartLen(value: string): number {
  let count = 0;
  while (count < value.length && /\s/u.test(value[count]!)) count += 1;
  return count;
}

export interface CharCursor {
  pos: number;
}

/** 对齐 Rust find_text_char_range：从 cursor 起查找 needle（含换行折叠重试），返回码点范围并推进 cursor */
export function findTextCharRange(
  fullText: string,
  needle: string,
  cursor: CharCursor,
): [number, number] | null {
  const trimmed = trimRust(needle);
  if (trimmed.length === 0) return null;
  const haystack = fullText.slice(cursor.pos);
  let found = haystack.indexOf(trimmed);
  if (found < 0) {
    found = haystack.indexOf(trimmed.split('\n').join(' '));
  }
  if (found < 0) return null;
  const start = cursor.pos + found;
  const end = start + trimmed.length;
  cursor.pos = end;
  return [charIndexAt(fullText, start), charIndexAt(fullText, end)];
}

/** UTF-16 索引之前（含）的 Unicode 码点数量（对齐 Rust char_index_at_byte 的字符语义） */
function charIndexAt(text: string, index: number): number {
  const bounded = Math.min(index, text.length);
  let count = 0;
  for (let i = 0; i < bounded; i += 1) {
    const code = text.charCodeAt(i);
    if (code < 0xd800 || code > 0xdbff) count += 1;
  }
  return count;
}
